<a href="https://colab.research.google.com/github/shoond/portfolio/blob/main/MSIS510_Case1_ClimateChange.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [74]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import statsmodels.formula.api as smf

# Climate Change

Study the relationship between avg global temp and several other factors. ClimateChange.csv contains data from May 1983 to December 2008.

# a
Split the data into training and test subsets (already done for us by professor)
* Training = observations up to and including 2006
* Test = observations after 2006

Build a linear regression model to predict Temp, using all other variables as predictors(except Year and Month). You should use the training set.


In [75]:
# Grab Climate change data

df_train_climate = pd.read_csv('https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTrain.csv')

df_test_climate = pd.read_csv('https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTest.csv')


# Confirm the training data is subset as instructed
print(df_train_climate['Year'].min())
print(df_train_climate['Year'].max())

#Confirm test data is subset as instructed
print(df_test_climate['Year'].min())
print(df_test_climate['Year'].max())


1983
2006
2007
2008


In [84]:
df_train_climate.head(5)    # Peek at training data

df_train_climate.describe() # Looking at count values to ensure every row has attr data

Year       Month         MEI         CO2          CH4  \
count   284.000000  284.000000  284.000000  284.000000   284.000000   
mean   1994.661972    6.556338    0.341923  361.414261  1745.841479   
std       6.845996    3.446768    0.929639   11.439691    45.669846   
min    1983.000000    1.000000   -1.586000  340.170000  1629.890000   
25%    1989.000000    4.000000   -0.323000  352.315000  1716.347500   
50%    1995.000000    7.000000    0.308500  359.890000  1758.605000   
75%    2001.000000   10.000000    0.898000  370.585000  1781.637500   
max    2006.000000   12.000000    3.001000  384.980000  1808.150000   

              N2O       CFC11       CFC12          TSI    Aerosols        Temp  
count  284.000000  284.000000  284.000000   284.000000  284.000000  284.000000  
mean   311.657225  252.487092  494.217546  1366.101437    0.017721    0.247799  
std      4.758513   20.987671   59.046642     0.401283    0.030014    0.181136  
min    303.677000  191.324000  350.113000  1365.426100    0.001600   -0.282000  
25%    307.657000  249.557750  462.543000  1365.754550    0.002700    0.118000  
50%    310.849500  260.373500  522.089000  1366.054500    0.006200    0.232500  
75%    316.129250  267.448000  540.972750  1366.399275    0.014000    0.406500  
max    320.451000  271.494000  543.813000  1367.316200    0.149400    0.739000

In [85]:
df_test_climate.head(5)     # Peek at test data

df_test_climate.describe()  # Looking at count values to ensure every row has attr data

Year      Month        MEI         CO2          CH4         N2O  \
count    24.000000  24.000000  24.000000   24.000000    24.000000   24.000000   
mean   2007.500000   6.500000  -0.509792  384.674583  1796.957083  321.084708   
std       0.510754   3.526299   0.640489    2.044560    11.398775    0.542589   
min    2007.000000   1.000000  -1.635000  380.900000  1771.890000  320.332000   
25%    2007.000000   3.750000  -1.043750  383.065000  1792.377500  320.557750   
50%    2007.500000   6.500000  -0.530500  384.525000  1798.045000  321.256000   
75%    2008.000000   9.250000  -0.036000  386.137500  1803.535000  321.408750   
max    2008.000000  12.000000   0.974000  388.500000  1814.180000  322.182000   

            CFC11       CFC12          TSI   Aerosols       Temp  
count   24.000000   24.000000    24.000000  24.000000  24.000000  
mean   245.890458  536.660417  1365.707737   0.004071   0.363000  
std      1.371970    1.427480     0.031728   0.000605   0.107847  
min    244.080000  534.906000  1365.657000   0.003100   0.074000  
25%    244.615500  535.114000  1365.675175   0.003600   0.307000  
50%    246.180500  536.964000  1365.714550   0.004100   0.380000  
75%    246.593000  537.377000  1365.726250   0.004500   0.414000  
max    248.372000  539.206000  1365.761600   0.005400   0.601000

# a_i
Build a linear regression model to predict Temp, using all other variables as predictors(except Year and Month). You should use the training set.

In [78]:
# Create loop to concat independent vars
columns = [col for col in df_train_climate.columns if col!='Temp' and col!='Year' and col!='Month']

joiner = '+'.join(columns)

col_all = 'Temp ~ ' + joiner

print(col_all)

Temp ~ MEI+CO2+CH4+N2O+CFC11+CFC12+TSI+Aerosols


In [79]:
# Fit linear reg model
Climate_linReg = smf.ols(col_all, data = df_train_climate)
Climate_linReg_results = Climate_linReg.fit()
print(Climate_linReg_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     103.6
Date:                Sun, 29 Oct 2023   Prob (F-statistic):           1.94e-78
Time:                        03:46:21   Log-Likelihood:                 280.10
No. Observations:                 284   AIC:                            -542.2
Df Residuals:                     275   BIC:                            -509.4
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -124.5943     19.887     -6.265      0.0

# a_ii
Evaluate the quality of the model.
* What is R2
> R2 = 0.751
* Which independent vars are significant?
> It seems like all independent variables are significant except CH4 and N2O.










# a_iii
Current scientific opinion is that N2O and CFC-11 are greenhouse gases. However,we see that regression coefficients for both N2O and CFC-11 are negative indicating these factors decrease global temps.

How do you explain this contradiction?
> As shown in the correlation_matrix below, there are high levels of multicollinearity amongst variables CO2, CH4, N2O, CFC11, and CFC12.

> Using multicollinear variables may result in numerical artifacts such as the contradiction described

In [80]:
correlation_matrix = df_train_climate.corr()
correlation_matrix.style.background_gradient(vmax = 1.0, vmin = 0.7)

# a_iv

Which independent variables is N2O highly correlated with (> 0.7)?
> N20 is highly correlated with: CO2, CH4 and CFC12

Which independent variables is CFC11 highly correlated with (> 0.7)?
> CFC11 is highly correlated with: CH4 and CFC12

# b

Build new linear regression model, this time only using **MEI, TSI, Aerosols, and N2O** as independent variables. Remember to use training set to build model.


In [81]:
# Create loop to concat independent vars
col_2 = [col for col in df_train_climate.columns if col=='MEI' or col=='TSI' or col=='Aerosols' or col=='N2O']

joiner = '+'.join(col_2)

col_2_all = 'Temp ~ ' + joiner

print(col_2_all)

Temp ~ MEI+N2O+TSI+Aerosols


In [82]:
# Fit linear reg model
Climate_linReg_2 = smf.ols(col_2_all, data = df_train_climate)
Climate_linReg_results_2 = Climate_linReg_2.fit()
print(Climate_linReg_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     103.6
Date:                Sun, 29 Oct 2023   Prob (F-statistic):           1.94e-78
Time:                        03:46:21   Log-Likelihood:                 280.10
No. Observations:                 284   AIC:                            -542.2
Df Residuals:                     275   BIC:                            -509.4
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -124.5943     19.887     -6.265      0.0

#b_i

How does the coefficient for N2O in this model compare to the previous model?

> N2O is positive compared to negative value of previous model indicating rise in N2O is a predictor of higher global temps.

#b_ii

How does the quality of this model compare to the previous one? Consider the R2 value and the significance of the independent vars when answering the question.

> R2 is slightly lower for this model, however all independent variables show as significant predictors of target variable, Temp.

# c

Using simple model from part b, calculate predictions for testing dataset.
* What is R2 on test set?
> 0.497
* What does this tell you about the model?
> The predictive variables we chose are roughly 50% probabilistic in predicting climate temperatures.


In [83]:
temp_pred = Climate_linReg_results_2.predict(df_test_climate)

temp_test = df_test_climate['Temp']

temp_train_mean = np.mean(df_train_climate['Temp'])

# R2 = (SSE/SST)
# Find SSE and SST first
# SSE = SUM((obs - pred)**2)
# SST = SUM((obs-mean(pred))**2)
SSE_temp = np.sum(((temp_test - temp_pred)**2))

SST_temp = np.sum(((temp_test - temp_train_mean)**2))

R2 = 1 - (SSE_temp / SST_temp)

print(R2)




0.4967794872206187
